### PR TITLE
Better typing for ClusterID and NetworkID

### PR DESCRIPTION
**Pull request by nmittler**
[ ] Configuration Infrastructure
[ ] Docs
[ ] Installation
[x] Networking
[ ] Performance and Scalability
[ ] Policies and Telemetry
[ ] Security
[ ] Test and Release
[ ] User Experience
[ ] Developer Infrastructure

Pull Request Attributes

Please check any characteristics that apply to this pull request. 

[ ] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Review by stevenctl — APPROVED** [review]
this could prevent easy mishaps – nice

Leaving a hold as well in case there are other opinions, seems low impact to me though.

**Comment by nmittler** [comment]
@stevenctl it also centralizes the `Equals` method. While this PR doesn't change any logic to use it, I know the hope was that all comparisons of `NetworkID` and `ClusterID` would use the same logic.

File: pilot/pkg/model/context.go [hunk 1]
```diff
@@ -643,14 +643,14 @@ func (node *Proxy) GetNetworkView() map[string]bool {
...
+func (node *Proxy) InNetwork(network NetworkID) bool {
+	return node == nil || SameOrEmpty(string(network), string(node.Metadata.Network))
 }
 
 // InCluster returns true if the proxy is in the given cluster, or if either
 // the proxy's cluster id or the given cluster id is unspecified ("").
-func (node *Proxy) InCluster(cluster string) bool {
-	return node == nil || SameOrEmpty(cluster, node.Metadata.ClusterID)
+func (node *Proxy) InCluster(cluster ClusterID) bool {
+	return node == nil || SameOrEmpty(string(cluster), string(node.Metadata.ClusterID))
```
Comment (lei-tang) [inline pilot/pkg/model/context.go:653]: Consider defining a String() for the type ClusterID to have a better control of how the type ClusterID is converted to string.

File: pilot/pkg/model/context.go [hunk 2]
```diff
@@ -643,14 +643,14 @@ func (node *Proxy) GetNetworkView() map[string]bool {
 
 // InNetwork returns true if the proxy is on the given network, or if either
 // the proxy's network or the given network is unspecified ("").
-func (node *Proxy) InNetwork(network string) bool {
-	return node == nil || SameOrEmpty(network, node.Metadata.Network)
+func (node *Proxy) InNetwork(network NetworkID) bool {
+	return node == nil || SameOrEmpty(string(network), string(node.Metadata.Network))
```
Comment (lei-tang) [inline pilot/pkg/model/context.go:647]: Consider defining a String() for the type NetworkID to have a better control of how the type NetworkID is converted to string.

**Comment by ramaraochavali** [comment]
LGTM

**Comment by nmittler** [comment]
@lei-tang PTAL

**Review by lei-tang — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Pull request merged**